### ISSUE TITLE
Discussion om Const Matrix and Const Tyle types

### ISSUE THREAD

**Issue by teonnik** [eth-cscs/DLA-Future#229]
Here I would like to open a discussion on what we are trying to get out of the Const/Non-Const variants of Matrix/Tile, namely `Matrix<const T, Device>` and `Tile<const T, Device>` (referred to as `ConstMatrix` and `ConstTile` from now on) as opposed to `Matrix<T, Device>` and `Tile<T, Device>` (referred to as `Matrix` and `Tile` from now on). 

The key question is what do we get from `ConstMatrix`/`ConstTile` that we don't already get from `const Matrix` and `const Tile`?

From the point of view of their memory layouts, they are essentially equivalent as the `non-const` variants inherit all the members of their `const` variants. From the perspective of exposed API, the `const` members in `ConstMatrix`/`ConstTile` can be merged into `Matrix`/`Tile` and `const` qualifier on them will have the same effect. From the point of view of interacting with HPX, when we pass a `hpx::shared_future<Tile>` to `hpx::dataflow`, the expected parameter in the functor passed to the dataflow is `const Tile&`. This is somewhat obscured at the moment as we use `auto&&` quite heavily. However, is there any meaningful difference between `const Tile&` and `const ConstTile &`?

Unless I am missing something, would we be open to merging ConstMatrix/ConstTile into Matrix/Tile to reduce the number of types we have to deal with?

**Comment by rasolca** [comment]
Unfortunately there is no way to get rid of the two different types.
The interface has been designed in this way after having faced many problems in the prototype, and mimics pointer interface and the difference between `const int*` and `int* const` (also `std::span` interface is similar).

This guarantee that:
- you can create a `ConstM/T` (but not a `M/T`) from a `const` pointer. (to interface with existing code)
- `ConstM/T` can be moved. (E.g. you can have a vector of `ConstM/T`)

However, having two types has a disadvantage, i.e. you cannot reference a `Tile` as a `ConstTile`,
but this has been solved deriving the non-const types from the const types.

`auto&&` can be used with future.get() (it returns a rvalue reference), but
it might have been used improperly in other cases.

In summary:
- you can have a `ConstM/T` reference of a `M/T`
- you can access, but not modify elements of `[const] ConstM/T`
- you can access and modify elements of `[const] M/T`
- you can move `M/T` and `ConstM/T`

Anyway there is no need to deal with multiple types e.g.:
if the elements are not be modified you use const ConstTile& (which works with Tile too),
otherwise you use Tile&& (or const Tile&)

**Comment by teonnik** [comment]
> The interface has been designed in this way after having faced many problems in the prototype, and mimics pointer interface and the difference between const int* and int* const (also std::span interface is similar).

Aha, thanks for pointing that out! :)

> auto&& can be used with future.get() (it returns a rvalue reference), but it might have been used improperly in other cases.

I didn't mean to say that it's used improperly in the code, just that it sometimes makes the type a little hard to deduce (i.e. obscures it).

> if the elements are not be modified you use const ConstTile& (which works with Tile too),
otherwise you use Tile&& (or const Tile&)

I missed the fact that `const Tile&` actually allows the elements to be modified, that makes it quite clear why `ConstTile` is needed.

I will close the issue now, thanks for clarifying!
